media/editor/dataDisplayContext.tsx

```diff
@@ -0,0 +1,265 @@
+import { EventEmitter, IDisposable } from "cockatiel";
+import { createContext, useContext, useEffect, useState } from "react";
+import { MessageType } from "../../shared/protocol";
+import { messageHandler } from "./state";
+import { Range } from "./util";
+
+export class FocusedElement {
+	public static readonly zero = new FocusedElement(false, 0);
+	public readonly key = BigInt(this.byte) << 1n | (this.char ? 1n : 0n);
+
+	constructor(
+		/** If true, the character rather than data is focused */
+		public readonly char: boolean,
+		/** Focused byte index */
+		public readonly byte: number,
+	) {}
+
+	/** Gets the other element at this byte (the character or non-character) */
+	public other(): FocusedElement {
+		return new FocusedElement(!this.char, this.byte);
+	}
+}
+
+/**
+ * Data management context component. Initially we used Recoil for this, but
+ * this ended up introducing performance issues with very many components.
```

Comment:
Things that use this aren't rerendered automatically, since the instance in the context doesn't change (it's all shallow equality). Instead this class exposes event listeners that are used in hooks (see the end of this file) which allow only the necessary components to be re-rendered. It's actually quite efficient 🙂